Added code to help DiHydrogen find cuDNN and CUB

bvanessen opened the pull request.

bvanessen committed:
Added code to help DiHydrogen find cuDNN and CUB

bvanessen committed:
Cleaning up dependencies on CUB and adding guards for when newer
versions of CUDA include CUB and it should be excluded.

bvanessen committed:
Changed Hydrogen to disable half support by default.

bvanessen committed:
Have LBANN force Hydrogen and DiHydrogen to build without half when the variant is disabled.

adamjstewart commented on var/spack/repos/builtin/packages/dihydrogen/package.py:
```diff
@@ -150,4 +150,13 @@ def cmake_args(self):
                         ' '.join(self.cuda_flags(cuda_arch))
                     ))
 
+        if '+cuda' in spec or '+legacy' in spec:
+            args.extend(['-DcuDNN_DIR={0}'.format(
+                spec['cudnn'].prefix)])
+
+        if spec.satisfies('^cuda@:10.99'):
+            if '+cuda' in spec or '+legacy' in spec:
+                args.extend(['-DCUB_DIR={0}'.format(
+                    spec['cub'].prefix)])
```
`list.append(...)` is easier than `list.extend([...])`

bvanessen replied:
I updated all of the spots where this is done.

adamjstewart reviewed (approved):
One minor request, otherwise looks good. Thanks for maintaining all of these packages!

bvanessen committed:
Added explicit variants to enusre that if LBANN is build without Cuda,
Aluminum, or Half support, it enforces those constraints for Hydrogen
and DiHydrogen.  Cleaned up the use of Python extend versus append in
LBANN and DiHydrogen recipes.

bvanessen committed:
Fixed Flake8
